src/search-and-replace-dialog.jsx

```diff
@@ -0,0 +1,276 @@
+                    count += Util.countOccurrences(particle.content, searchString);
+                }
+            }
+        }
+
+        return count;
+    },
+
+    handleNextSearchResult() {
+        var searchIndex = this.props.searchIndex;
+        searchIndex ++;
+        searchIndex = searchIndex % this.state.searchResultCount;
+
+        this.props.onChange({ searchIndex });
+    },
+
+    handlePreviousSearchResult() {
+        var searchIndex = this.props.searchIndex;
+        searchIndex --;
+        if (searchIndex < 0) {
+            searchIndex = this.state.searchResultCount - 1;
+        }
+
+        this.props.onChange({ searchIndex });
+    },
+
+    handleReplaceAll() {
+        var replaceString = this.state.replaceString;
+        var { searchString, question, hints, particle } = this.props;
+
+        var regex = new RegExp(Util.escapeRegExp(searchString), "g");
+
+        var replaceFunc = function(obj) {
+            if (obj) {
+                if (Array.isArray(obj)) {
+                    obj.forEach(replaceFunc);
+                } else {
+                    var indices = Util.getIndicesOf(obj.content, searchString);
+                    obj.content = obj.content.replace(regex, (match, offset) => {
+                        if (indices.indexOf(offset) !== -1) {
+                            return replaceString
+                        } else {
+                            return match;
+                        }
+                    });
+                }
+            }
+        };
+
+        replaceFunc(question);
+        replaceFunc(hints);
+        replaceFunc(particle);
+
+        // handle the case where the replaceString contains one or more copies
+        // of searchString
+        // TODO(kevinb7) this is quite right
+        var searchResultCount = this.getSearchResultCount(searchString);
+
+        this.setState({ searchResultCount });
+        this.props.onChange({ question, hints, json: particle, searchIndex: 0 });
+    },
+
+    handleReplace() {
+        var replaceString = this.state.replaceString;
+        var { searchIndex, searchString, question, hints, particle } = this.props;
+
+        var regex = new RegExp(searchString, "g");
+        var replaced = false;
+        var globalIndex = 0;
+
+        var replaceFunc = function(obj) {
+            if (!replaced && obj) {
+                if (Array.isArray(obj)) {
+                    obj.forEach(replaceFunc);
+                } else {
+                    var indices = Util.getIndicesOf(obj.content, searchString);
+                    obj.content = obj.content.replace(regex, (match, offset) => {
+                        if (indices.indexOf(offset) !== -1) {
+                            if (!replaced && globalIndex === searchIndex) {
+                                replaced = true;
+                                globalIndex++;
+                                return replaceString;
+                            } else {
```

Comment:
We might need to change some of this to avoid mutating props, so let's see what this looks like after that, then see if we can simplify it or add some helpful comments :)